### PR TITLE
Types

### PR DESCRIPTION
**Pull request by SimonGuilloud**
This PR is about making a front end for first order logic, allowing to define more specialized Terms, starting with functions, numbers and pairs and ultimately arbitrary types. The implementation aims to be as general as possible to allow for future development.
Incidentally, it solves some unsatisfactory problems of using the Kernel's FOL: The difference between Variables and VariableLabels, runtime assertions to check arity (with the updated "front" this is checked at compile time), etc.

With FOL redefined, it needs to be integrated to the existing proof framework, while maintaining correspondance with kernel objects. Depending of the kind of object, this range from easy to systematic but tedious to non-trivial.

### CODE REVIEW

**Comment by SimonGuilloud** [comment]
This should be ready for merge.

File: lisa-utils/src/main/scala/lisa/utils/K.scala [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package lisa.utils
+
+object K {
```
Comment (sankalpgambhir) [inline lisa-utils/src/main/scala/lisa/utils/K.scala:3]: I agree that it's nice to be able to use it as K, but it's probably better to name it `Kernel` and alias it after, even a `val K = Kernel` right next to it is fine

File: project/build.properties [hunk 2]
```diff
@@ -1 +1 @@
-sbt.version = 1.5.7
+sbt.version = 1.8.2
```
Comment (sankalpgambhir) [inline project/build.properties:1]: Can we have a newline at the end for consistency?

**Comment by sankalpgambhir** [comment]
Looks great overall; I checked as much of the technical stuff as I could at this moment. All theory files pass as well! Minor comments as above ^^^ and then we can merge!

File: lisa-examples/src/main/scala/Example.scala
Comment (sankalpgambhir) [inline lisa-examples/src/main/scala/Example.scala:1]: `Example.scala` needs `scalafix; scalafmt` run on it, you'll have to switch to the project and do it separately

Comment (SimonGuilloud) [reply lisa-examples/src/main/scala/Example.scala:1]: ok !

File: project/build.properties (hunk 2 above)
Comment (SimonGuilloud) [reply project/build.properties:1]: not sure what you mean

File: lisa-utils/src/main/scala/lisa/utils/K.scala (hunk 1 above)
Comment (SimonGuilloud) [reply lisa-utils/src/main/scala/lisa/utils/K.scala:3]: This is already supposed to be the alias :)

**Review by sankalpgambhir — APPROVED** [review]
Looks good now! Thanks for changing according to comments

**Pull request merged**